src/framework/components/element/pointer-events.js

```diff
@@ -0,0 +1,190 @@
+            for(var i = 0; i < this.entity.children.length; i++) {
+                var element =  this.entity.children[i];
+                if (element && element.element) {
+                    element.element[ name ].apply( element.element, eventData );
+                }
+            }
+        },
+
+        // Handles "down" pointer event – might be coming from touch or
+        // a mouse.
+        _pointerEventDown: function(point) {
+            point = this._parentPointToLocalPoint(point);
+
+            if (!this._testPointerEvent(point)) {
+                return;
+            }
+
+            this._passPointerEventToChildren("_pointerEventDown", [ point ]);
+            this.fire(pc.POINTEREVENT_DOWN, point);
+        },
+
+        // Handles "up" pointer event – might be coming from touch or
+        // a mouse.
+        _pointerEventUp: function(point) {
+            point = this._parentPointToLocalPoint(point);
+
+            if (!this._testPointerEvent(point)) {
+                return;
+            }
+
+            this._passPointerEventToChildren("_pointerEventUp", [ point ]);
+            this.fire(pc.POINTEREVENT_CLICK, point);
+            this.fire(pc.POINTEREVENT_UP, point);
+        },
+
+        // Fires pointer leave event and also makes all children do so.
+        _ensurePointerLeaveEvent: function(point) {
+            this._pointerOver = false;
+            this.fire(pc.POINTEREVENT_LEAVE, point); 
+
+            this._passPointerEventToChildren( "_ensurePointerLeaveEvent", [ point ]);
+        },
+
+        // Handles "move" pointer event – might be coming from touch or
+        // a mouse.
+        _pointerEventMove: function(point) {
+            point = this._parentPointToLocalPoint(point);
+
+            if (!this._testPointerEvent(point)) {
+                if (this._pointerOver) {
+                    this._ensurePointerLeaveEvent(point);            
+                }
+
+                return;
+            }
+
+            this._passPointerEventToChildren("_pointerEventMove", [ point ]);
+
+            if (!this._pointerOver) {
+                this._pointerOver = true;
+                this.fire(pc.POINTEREVENT_ENTER, point);                
+            }
+
+            this.fire(pc.POINTEREVENT_MOVE, point);
+        },
+
+        // Handles "scroll" pointer event – might be coming from touch or
+        // a mouse.
+        _pointerEventScroll: function(point, amount) {
+            point = this._parentPointToLocalPoint(point);
+
+            if (!this._testPointerEvent(point)) {
+                return;
+            }
+
+            this._passPointerEventToChildren("_pointerEventScroll", [ point, amount ]);
+            this.fire(pc.POINTEREVENT_SCROLL, point, amount);
+        },
+
+        // Mouse-specific event handler.
+        _onMouseDown: function(mouseEvent) {
+            this._pointerEventDown( new pc.Vec3( mouseEvent.x, mouseEvent.y, 0 ) );
```

Comment:
Allocation, is no good, consider another approach.
In sense of providing object as argument in event, it can be clearly stated that this object should only be used by developer within sync life of a function.
Additionally, what about mouse button that is used during mouse event, like right mouse, or mouse wheel clicked, etc? :) Providing second argument - original event, would be not bad too.
There is `Vec2` for doing 2D vectors. And in this case, using X, Y simply numbers would not be bad either.
To challenge this even further, why not to stick to similar API as current mouse event handling does? Where it provides mouse event object.